src/clients/AcrossHubPoolClient.ts

```diff
@@ -0,0 +1,667 @@
+  }
+
+  // This should find the ProposeRootBundle event whose bundle block number for `chain` is closest to the `block`
+  // without being smaller. It returns the bundle block number for the chain or undefined if not matched.
+  getRootBundleEvalBlockNumberContainingBlock(
+    latestMainnetBlock: number,
+    block: number,
+    chain: number,
+    chainIdList: number[]
+  ): number | undefined {
+    let endingBlockNumber: number | undefined;
+    // Search proposed root bundles in reverse chronological order.
+    for (let i = this.proposedRootBundles.length - 1; i >= 0; i--) {
+      const rootBundle = this.proposedRootBundles[i];
+      const nextRootBundle = this.getFollowingRootBundle(rootBundle);
+      if (!this.isRootBundleValid(rootBundle, nextRootBundle ? nextRootBundle.blockNumber : latestMainnetBlock)) {
+        continue;
+      }
+
+      // 0 is the default value bundleEvalBlockNumber.
+      const bundleEvalBlockNumber = this.getBundleEndBlockForChain(
+        rootBundle as ProposedRootBundle,
+        chain,
+        chainIdList
+      );
+
+      // Since we're iterating from newest to oldest, bundleEvalBlockNumber is only decreasing, and if the
+      // bundleEvalBlockNumber is smaller than the target block, then we should return the last set `endingBlockNumber`.
+      if (bundleEvalBlockNumber <= block) {
+        if (bundleEvalBlockNumber === block) {
+          endingBlockNumber = bundleEvalBlockNumber;
+        }
+        break;
+      }
+      endingBlockNumber = bundleEvalBlockNumber;
+    }
+    return endingBlockNumber;
+  }
+
+  // TODO: This might not be necessary since the cumulative root bundle count doesn't grow fast enough, but consider
+  // using _.findLast/_.find instead of resorting the arrays if these functions begin to take a lot time.
+  getProposedRootBundlesInBlockRange(startingBlock: number, endingBlock: number): ProposedRootBundle[] {
+    return this.proposedRootBundles.filter(
+      (bundle: ProposedRootBundle) => bundle.blockNumber >= startingBlock && bundle.blockNumber <= endingBlock
+    );
+  }
+
+  getCancelledRootBundlesInBlockRange(startingBlock: number, endingBlock: number): CancelledRootBundle[] {
+    return sortEventsDescending(this.canceledRootBundles).filter(
+      (bundle: CancelledRootBundle) => bundle.blockNumber >= startingBlock && bundle.blockNumber <= endingBlock
+    );
+  }
+
+  getDisputedRootBundlesInBlockRange(startingBlock: number, endingBlock: number): DisputedRootBundle[] {
+    return sortEventsDescending(this.disputedRootBundles).filter(
+      (bundle: DisputedRootBundle) => bundle.blockNumber >= startingBlock && bundle.blockNumber <= endingBlock
+    );
+  }
+
+  getLatestProposedRootBundle(): ProposedRootBundle {
+    return this.proposedRootBundles[this.proposedRootBundles.length - 1] as ProposedRootBundle;
+  }
+
+  getFollowingRootBundle(currentRootBundle: ProposedRootBundle): ProposedRootBundle | undefined {
```

Comment:
Set the return to include undefined to remove the error on the return undefined this this function